### PR TITLE
Fix version-file truncation

### LINKED ISSUE #42: We don't truncate the version.json file when we write it
It's common to end up with a version file that contains junk from its previous incarnation at the end, because we don't truncate the file when we open it for write.

### PR DESCRIPTION
**Pull request by Smaug123**
Fixes #42.

(The version history is meaningful, if you want to review commit-by-commit.)

### CODE REVIEW

File: ApiSurface/ApiSurface.fs [hunk 1]
```diff
@@ -205,7 +205,7 @@ module ApiSurface =
                 Version = sprintf "%d.%d" version.Major version.Minor
             }
 
-        use writer = versionFile.OpenWrite ()
+        use writer = versionFile.Open FileMode.Create
```
Comment (Smaug123) [inline ApiSurface/ApiSurface.fs:208]: This is the only operative line of the PR. The rest is tests.

File: ApiSurface/Test/TestVersionFile.fs [hunk 2]
```diff
@@ -120,3 +122,48 @@ module TestVersionFile =
...
+                }
+        }
+
+    let private incrementVersion (version : string) =
+        match version.Split '.' with
+        | [| major ; _minor |] -> sprintf "%i.0" (int major + 1)
+        | _ -> failwithf "Unrecognised version string: %s" version
+
+    [<Test>]
+    let ``version JSON file can be written`` () =
```
Comment (kojo12228) [inline ApiSurface/Test/TestVersionFile.fs:149]: The name similarity to the test confused me for a second.

More importantly, I couldn't quite figure out what was wrong with the test in it's present form (and I don't have .NET 3.1 installed, but I locally changed the test project to only target .NET 7). It is just that the generated version data vs the update `ApiSurface.updateVersionJson` could potentially lead to lines from the previous write remaining, creating invalid JSON?

**Review by kojo12228 — COMMENTED** [review]
The change makes sense, but I don't quite understand the test currently

File: ApiSurface/Test/TestVersionFile.fs (hunk 2 above)
Comment (Smaug123) [reply ApiSurface/Test/TestVersionFile.fs:149]: Yes. The file is not truncated before writing, on `main`, so old parts of the end of the file may remain after we've overwritten it.

File: ApiSurface/Test/RedirectOutput.fs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    let writer = new StreamWriter (substitute)
+
+    do
+        let isLocked = Interlocked.Increment (RedirectOutput.IsLocked : int ref)
+
+        if isLocked = 1 then
+            original <- Console.Out
+            Console.SetOut writer
+        else
+            failwith "Attempted to use RedirectOutput twice in parallel"
```
[the remark this answers is not in the document]
Comment (Smaug123) [reply ApiSurface/Test/RedirectOutput.fs:20]: Don't understand the question, sorry!

[the remark this answers is not in the document]
Comment (kojo12228) [reply ApiSurface/Test/RedirectOutput.fs:20]: I was being cryptic, I'm suggesting the message should be _Attempted to use RedirectOutput more than once in parallel_

[the remark this answers is not in the document]
Comment (Smaug123) [reply ApiSurface/Test/RedirectOutput.fs:20]: If someone tried to use it more than once, they necessarily tried to use it twice!

**Review by nojaf — APPROVED** [review]

**Pull request merged**